File: langs/typescript/src/services/nullDispatcher.ts
/**
 * NullDispatcher — null-object variant of IDispatcher.
 *
 * See spec/11-threading.md §"Null variant" and ADR-0017.
 *
 * Both schedulers are queueScheduler, which executes scheduled work
 * synchronously on the calling thread (no microtask, no macrotask hop).
 */
import { queueScheduler, type SchedulerLike } from "rxjs";
import type { IDispatcher } from "./dispatcher.js";

export class NullDispatcher implements IDispatcher {
  /** Shared singleton instance. */
  static readonly INSTANCE: NullDispatcher = new NullDispatcher();

  private constructor() {}

  readonly foreground: SchedulerLike = queueScheduler;
  readonly background: SchedulerLike = queueScheduler;
}

File: langs/swift/Sources/VMx/Services/NullDispatcher.swift
//
// NullDispatcher — null-object variant of `Dispatcher`.
//
// See spec/11-threading.md §"Null variant" and ADR-0017. Both targets
// run synchronously on the calling thread, matching `NullDispatcher`
// behavior in the other flavors.
//
import Foundation

public final class NullDispatcher: Dispatcher {
    /// Shared singleton instance.
    public static let INSTANCE = NullDispatcher()

    private init() {}

    public func scheduleForeground(_ work: @escaping () -> Void) {
        work()
    }

    public func scheduleBackground(_ work: @escaping () -> Void) {
        work()
    }
}
